core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -345,73 +343,69 @@ class TransactionStateManager(brokerId: Int,
+            info(s"Appending transaction message $newMetadata for $transactionalId failed due to " +
               s"${status.error.exceptionName}, returning ${Errors.NOT_COORDINATOR} to the client")
 
             Errors.NOT_COORDINATOR
 
           case Errors.MESSAGE_TOO_LARGE
                | Errors.RECORD_LIST_TOO_LARGE =>
 
-            error(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+            error(s"Appending transaction message $newMetadata for $transactionalId failed due to " +
               s"${status.error.exceptionName}, returning UNKNOWN error code to the client")
 
             Errors.UNKNOWN
 
           case other =>
-            error(s"Appending metadata message $txnMetadata for $transactionalId failed due to " +
+            error(s"Appending metadata message $newMetadata for $transactionalId failed due to " +
               s"unexpected error: ${status.error.message}")
 
             other
         }
       }
 
       if (responseError == Errors.NONE) {
-        def completeStateTransition(metadata: TransactionMetadata, newState: TransactionState): Boolean = {
-          // there is no transition in this case
-          if (metadata.state == Empty && newState == Empty)
-            true
-          else
-            metadata.completeTransitionTo(txnMetadata.state)
-        }
         // now try to update the cache: we need to update the status in-place instead of
         // overwriting the whole object to ensure synchronization
-          getTransactionState(transactionalId) match {
-            case Some(metadata) =>
-              metadata synchronized {
-                if (metadata.pid == txnMetadata.pid &&
-                  metadata.producerEpoch == txnMetadata.producerEpoch &&
-                  metadata.txnTimeoutMs == txnMetadata.txnTimeoutMs &&
-                  completeStateTransition(metadata, txnMetadata.state)) {
-                  // only topic-partition lists could possibly change (state should have transited in the above condition)
-                  metadata.addPartitions(txnMetadata.topicPartitions.toSet)
-                } else {
-                  throw new IllegalStateException(s"Completing transaction state transition to $txnMetadata while its current state is $metadata.")
-                }
+        getTransactionState(transactionalId) match {
+          case Some(epochAndMetadata) =>
+            val metadata = epochAndMetadata.transactionMetadata
+
+            metadata synchronized {
+              if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch && metadata.completeTransitionTo(newMetadata)) {
+                debug(s"Updating $transactionalId's transaction state to $newMetadata with coordinator epoch $coordinatorEpoch for $transactionalId succeeded")
+              } else {
+                // the cache may have been changed due to txn topic partition emigration and immigration,
```

Comment:
Good point, we could consider throw exception directly instead of returning false.

@junrao @hachikuji There are a few of those suggestions that we can do in a follow-up PR. I have marked those places with `TODO`s.